### ISSUE TITLE
TypeScript documentation

### ISSUE THREAD

**Issue by aomarks** [Polymer/docs#2439]
Since we have increasingly good TypeScript support for Polymer, how about we add a dedicated page to the documentation site about it? This could include:

- How to load the official typings we are shipping starting with Polymer 2.4 .(https://github.com/Polymer/polymer/tree/master/types)
- How to use the Polymer decorators (https://github.com/Polymer/polymer-decorators).
- How to generate typings for your own Closure-style elements (https://github.com/Polymer/gen-typescript-declarations).
- How to set up your TypeScript project to work with Polymer build etc. (needs investigation).

Happy to take a first stab at it.

**Comment by abdonrd** [comment]
A Polymer CLI template with TypeScript would be very good.

**Comment by aomarks** [comment]
Note to self, include explanation of how to use `mixinBehaviors`, as discussed here: https://github.com/Polymer/polymer-modulizer/issues/133

**Comment by arthurevans** [comment]
SGTM.

**Comment by kito99** [comment]
@abdonrd my goal is to eventually do that with https://github.com/virtua-tech/polymer-typescript-starter-kit; right now you have to clone it. Still working on the finishing touches, but it's a good start.

**Comment by merlinnot** [comment]
Ever since `static get template` landed, my `html` files contain literally two things: html imports and one script tag. I’m just wondering... did any of you guys know a better solution?

**Comment by aomarks** [comment]
> Ever since static get template landed, my html files contain literally two things: html imports and one script tag. I’m just wondering... did any of you guys know a better solution?

This isn't related to this issue, maybe ask in our slack?

**Comment by merlinnot** [comment]
I thought it might fit into “setting up TypeScript” project :)

We must consider what should we recommend in the docs, since it might look a little weird at a first glance. That’s why I asked if anyone knows a better solution, so it can be documented. Current approach have multiple disadvantages, both in development and in production.

**Comment by kito99** [comment]
@merlinnot isn't this a Polymer 3 question? I'm assuming the goal is get TypeScript docs up for Polymer 2.x first.

**Comment by merlinnot** [comment]
No, it’s not. It’s now possible to get rid of almost everything from html files in Polymer 2 (since 2.4), despite those two things I’ve mentioned: imports and a script tag to load actual js.

It’s a bit annoying for a few reasons:
- you have to keep in sync imports in `.html` files and type references in `.ts` files.
- the request chain is unnecessarily longer than it could be
- no tool will analyze this in terms of jumping to definition etc.
- it literally doubles number of files to download and maintain ;)

I’m fully aware of an existence of modulizer and how it’s supposed to smooth the transition to Polymer 3, I really keep an eye on everything what’s going on. I literally watch every single repository in Polymer, PolymerLabs and PolymerElements repository and read all of the PRs and issues carefully multiple times a day.

I’m just wondering if it would be possible to come up with a better solution before the current one (the best I know for TypeScript) will get documented. That’s why I started this discussion here, not on slack nor in a separate issue.

**Comment by merlinnot** [comment]
Some time has to pass by until we’ll be able to transition large codebases relying on multiple components to Polymer 3. I’d love that, trust me. It’s just not that easy.

I’m currently in the process of dropping more and more external dependencies in projects I contribute to so they can be transitioned more easily to Polymer 3 (no need to wait for those to be updated), but at the present time you just can’t go with Polymer 3 to production. Let’s be real :)

Take a look at PolymerFire, sooo many projects rely on those elements and it’s now abandoned... no chance to get it ready for Polymer 3.

**Comment by merlinnot** [comment]
@aomarks, I’d love to help you out with this, just let me know if you’d like to split up the work or need some expertise.
